### PR TITLE
add HaarTransform

### PR DESCRIPTION
**Pull request by martinjankowiak**
this discrete transformation can be useful in the context of time series models, where it can improve mixing in HMC inference.

this implementation is based on a recursive formulation suggested by @fritzo and @fehiepsi.

### CODE REVIEW

File: pyro/distributions/transforms/haar.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    :param int dim: Dimension along which to transform. Must be negative.
+        This is an absolute dim counting from the right.
+    :param bool flip: Whether to flip the time axis before applying the
+        Haar transform. Defaults to false.
+    """
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+
+    def __init__(self, dim=-1, flip=False, cache_size=0):
```
Comment (martinjankowiak) [inline pyro/distributions/transforms/haar.py:30]: do i need the cache stuff?

File: tests/distributions/test_haar.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+from pyro.distributions.transforms import HaarTransform
+from tests.common import assert_equal
+
+
+@pytest.mark.parametrize('size', [1, 3, 4, 7, 8, 9])
+def test_haar_ortho(size):
+    haar = HaarTransform()
+    eye = torch.eye(size)
+    mat = haar(eye)
+    assert_equal(eye, mat @ mat.t())
```
Comment (martinjankowiak) [inline tests/distributions/test_haar.py:13]: test for orthogonality

File: tests/distributions/test_transforms.py [hunk 3]
```diff
@@ -192,6 +192,11 @@ def test_discrete_cosine(self):
         self._test(lambda input_dim: T.DiscreteCosineTransform(smooth=1.0))
         self._test(lambda input_dim: T.DiscreteCosineTransform(smooth=2.0))
 
+    def test_haar_transform(self):
```
Comment (martinjankowiak) [inline tests/distributions/test_transforms.py:195]: this should check shapes, inverses, and the log det jacobian

File: pyro/distributions/transforms/haar.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        assert isinstance(dim, int) and dim < 0
+        self.event_dim = -dim
+        self.flip = flip
+        super().__init__(cache_size=cache_size)
+
+    def __eq__(self, other):
```
Comment (martinjankowiak) [inline pyro/distributions/transforms/haar.py:36]: do we need the custom `__eq__`?

File: pyro/ops/tensor_utils.py [hunk 5]
```diff
@@ -6,6 +6,9 @@
 import torch
 
 
+root_two_inverse = 1.0 / math.sqrt(2.0)
```
[the remark this answers is not in the document]
Comment (martinjankowiak) [reply pyro/ops/tensor_utils.py:9]: done

**Comment by martinjankowiak** [comment]
@fritzo addressed comments. 

why don't you go ahead and lump in with `UnitJacobianReparam`

File: pyro/distributions/transforms/haar.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    is equivalent to a block-structured Haar transform in which block
+    sizes decrease by factors of one half from left to right.
+
+    :param int dim: Dimension along which to transform. Must be negative.
+        This is an absolute dim counting from the right.
+    :param bool flip: Whether to flip the time axis before applying the
+        Haar transform. Defaults to false.
+    """
+    domain = constraints.real
+    codomain = constraints.real
```
Comment (fehiepsi) [inline pyro/distributions/transforms/haar.py:27]: I think those are real_vector

File: pyro/distributions/transforms/haar.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            y = y.transpose(dim, -1)
+        x = inverse_haar_transform(y)
+        if self.flip:
+            x = x.flip(-1)
+        if dim != -1:
+            x = x.transpose(dim, -1)
+        return x
+
+    def log_abs_det_jacobian(self, x, y):
+        return x.new_zeros((1,) * self.event_dim)
```
Comment (fehiepsi) [inline pyro/distributions/transforms/haar.py:63]: I think the shape is x.shape[:-self.event_dim]

Comment (fritzo) [reply pyro/distributions/transforms/haar.py:63]: yeah I think this is also wrong in `DiscreteCosineTransform` from where Martin copied this. @martinjankowiak would you mind fixing it there too?

Comment (martinjankowiak) [reply pyro/distributions/transforms/haar.py:63]: what if `event_dim` is in the middle? doesn't this need to be a sum of 2 tuples?

Comment (fritzo) [reply pyro/distributions/transforms/haar.py]: the return value should just have `shape = batch_shape`

**Pull request merged**